src/compiler/nvcc.rs

```diff
@@ -516,58 +544,82 @@ impl CompileCommandImpl for NvccCompileCommand {
-                            .filter_map(|path| path.ok())
-                            .filter_map(|path| {
-                                path.file_name()
-                                    .to_str()
-                                    .map(|file| (path.path(), file.to_owned()))
-                            })
-                            .try_fold((), |res, (path, file)| fs::rename(path, dst.join(file)))
-                    })
-                    .ok()
-            });
+        let (mut nvcc_internal_files, nvcc_subcommand_groups) =
+            group_nvcc_subcommands_by_compilation_stage(
+                creator,
+                executable,
+                arguments,
+                compile_flag,
+                cwd,
+                out_dir.as_path(),
+                keep_dir.clone(),
+                env_vars,
+                host_compiler,
+                output_path,
+            )
+            .await?;
+
+        let mut maybe_keep_temps_then_clean = || {
+            // Move and/or delete nvcc's internal files.
+            //
+            // If the caller passed `-keep` or `-keep-dir`, copy the internal
+            // files to the requested location. We do this because we override
+            // `-keep` and `-keep-dir` in our `nvcc --dryrun` call.
+            //
+            // Renames the files back to the original names nvcc gave them.
+            let temps_kept = {
+                if let Some(keep_dir) = keep_dir.as_ref() {
+                    // Ensure `--keep-dir` exists
+                    fs::create_dir_all(keep_dir)?;
+
+                    // Rename files back to original nvcc names
+                    for (curr, prev) in nvcc_internal_files.drain() {
+                        let src = out_dir.join(curr);
+                        let dst = out_dir.join(prev);
+                        if dst == src || !src.exists() {
+                            continue;
+                        }
+                        trace!(
+                            "[{}]: maybe_keep_temps_then_clean rename {src:?} -> {dst:?}",
+                            output_path.display(),
+                        );
+                        fs::rename(src, dst)?;
+                    }
+
+                    // Move intermediate files to `--keep-dir`
+                    if let Ok(entries) = fs::read_dir(out_dir) {
+                        for entry in entries {
+                            let entry = entry?;
+                            let src = entry.path();
+                            let dst = keep_dir.join(entry.file_name());
+                            if dst == src || !src.exists() {
+                                continue;
+                            }
+                            trace!(
+                                "[{}]: maybe_keep_temps_then_clean move {src:?} -> {dst:?}",
+                                output_path.display()
+                            );
+                            fs::rename(src, dst)?;
```

Comment:
Can we ensure it's the same filesystem? Or have a fallback? Otherwise `rename` will fail.

Reply:
How would we check if `src` and `dst` are on the same file system? Should I use `fs::copy()` and `fs::remove()` instead of `fs::rename()`?